machinery/core_types.go

```diff
@@ -0,0 +1,86 @@
+package machinery
+
+import (
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+// These are wrappers for Core API types so instances can be used as targetables in the topology.
+
+type Namespace struct {
+	*core.Namespace
+
+	attachedPolicies []Policy
+}
+
+var _ Targetable = &Namespace{}
+
+func (n *Namespace) GetURL() string {
+	return UrlFromObject(n)
+}
+
+func (n *Namespace) SetPolicies(policies []Policy) {
+	n.attachedPolicies = policies
+}
+
+func (n *Namespace) Policies() []Policy {
+	return n.attachedPolicies
+}
```

Comment:
I guess `Namespace` is a special case, which we need to think in the context of the hierarchy of network resources.

This is a targetable that every other targetable the lib provides a wrapper for (with the exception of GatewayClass) can be linked to. The link itself is pretty straightforward and probably we could provide the helpers. But then we need to think what a link means in terms of inheritance.

E.g. a HTTPRoute is linked to a Gateway (via `spec.parentRef`) and to a Namespace (via `metadata.namespace`). In a scenario where we have policies attached to each of these resources, the final effective policy at the level of that one HTTPRoute will be composed out of `Gateway policies → HTTPRoute Namespace policies → HTTPRoute policies`. However, for an occasional second HTTPRoute in the same namespace, it doesn't mean necessarily the Gateway policies will have any effect, because it also depends if the second HTTPRoute is parented by that Gateway or not. Furthermore, the Gateway itself could be in the same namespace of the routes. I wonder what happens to the hierarchy of policies in this case.

One idea I had was of defining wrappers for namespace for each kind of namespaced targetable, i.e. `GatewayNamespace`, `HTTPRouteNamespace`, etc. But we would not be able to resolve the Policy → Namespace links from the targetRef only. So we need to think this through again.

Reply:
Yes, exactly, `Namespace` looks to be a special case as a `Targetable` and what it means in terms of inheritance for the effective policy. For now, I was thinking of just adding a wrapper for this as part of this PR. The link functions can be provided as a seperate PR after this is thought through

Reply:
I've added a linking functions for `Namespace -> Gateway ` and  `Namespace -> xRoute` that looks like this at the moment:

<img width="1456" alt="image" src="https://github.com/user-attachments/assets/4f2a6c45-7777-474b-b546-5b0c5b852c25">

If a namespace contains a route but parent refs a gateway in a different namespace, there is a direct link the the route to the namespace. Otherwise if the route parent ref gateway is in the same namespace as the route, this link from `Namespace -> Route` is skipped as it is already linked by `Namespace -> Gateway -> Route`.

Let me know your thought on this

Reply:
> if the route parent ref gateway is in the same namespace as the route, this link from `Namespace -> Route` is skipped as it is already linked by `Namespace -> Gateway -> Route`.

I understand this would make for a clean graph, but it seems to add complexity IMO. I guess I would keep it simple: if resource _R_ is in namespace _N_, then `N → R`.

Moreover, conceptually the relationship between namespace and route is not transitive via the gateway, but a direct one. It exists regardless of the existence of the gateway. Users should get the route _and_ gateway when asking for the child targetables of a namespace if these resources are both in the namespace.

IDK... to me it looks like we're better off without the special case. But if you've put some thought on this. There may be something I'm missing perhaps.

Reply:
We've decided to move Namespace linking to a separate PR

Reply:
If the Namespace linking is in a a separate PR, should the adding of the wrapper also be in that PR?

Reply:
I think it's fine / safe to keep the wrapper in this PR as it was an original ask of the issue

Reply:
I am wondering is this not effectively dead code currently. I know it is safe to have it there but what use does it give a user of the library. I am looking at this from the point it is harder to remove code than add it in later. If it is to be left there, sure I am okay with that but if there is no use case of it today I think we should remove it.

Reply:
I guess the only use now by having the wrapper is that users of this library can just use this struct while providing their own link functions if they so choose (until we provide the link functions, but even then, this is all opt-in depending on how they want to create their topology). Without this, a user would also need to wrap the `Namespace` struct themselves with pretty much the same implementation (unless they want to implement it differently) 🤔 

I've no strong preference of this either, but if we do decide to remove this until later, I'll also remove the `ReferenceGrant` wrapper, since the same will apply to it

Reply:
> users of this library can just use this struct while providing their own link functions

Agreed. This may also be a good way to learn the best way to provide the built-in link funcs later on.